File: lib/to_mdx/parsers/open_document/meta/user_defined.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Meta
      class UserDefined < Base
        attr_reader :property_name, :property_value

        def initialize(element = nil, context = {})
          super(element, context)
          @property_name = element&.attributes&.[]("meta:name")
          @property_value = process_property_value(element&.text&.strip)
        end

        def to_mdx
          ""
        end

        def to_s
          return "" unless property_name && property_value
          "#{property_name}: #{property_value}"
        end

        private

        def process_property_value(value)
          return nil unless property_name
          return nil unless value && !value.empty?

          underscored_name = property_name.downcase.gsub(/[^a-z0-9_]+/, " ").gsub(/\s+/, "_")

          case underscored_name
          when "keywords", "tags"
            value = value.split(", ").map(&:strip)
            unless value.empty?
              context["keywords"] ||= []
              context["keywords"] << value
            end
            value
          when /^(event|company)_(.+)$/
            namespace, property_name = $1, $2
            context[namespace] ||= {}
            context[namespace][property_name] = value
            context[namespace]
          when "date", "presentation_date"
            presentation_date = value.is_a?(Date) ? value : Date.parse(value)
            context["date"] = presentation_date
            presentation_date
          else
            context[underscored_name] = value
            value
          end
        end

        register_for "meta:user-defined"
      end
    end
  end
end

File: src/parsers/open-document/meta/user_defined.ts
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class UserDefined extends BaseElement {
  private name: string | null;
  private value: unknown;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.name = this.attr("meta:name");
    this.value = this.processValue((element.textContent || "").trim(), context);
  }

  toMdx(): string {
    return "";
  }

  toString(): string {
    if (!this.name || !this.value) {
      return "";
    }
    // eslint-disable-next-line @typescript-eslint/no-base-to-string
    return `${this.name}: ${this.value}`;
  }

  private processValue(value: string, context: ParseContext): unknown {
    if (!this.name || !value) {
      return null;
    }

    const underscoredName = this.name
      .toLowerCase()
      .replace(/[^a-z0-9_]+/g, " ")
      .replace(/\s+/g, "_");

    switch (underscoredName) {
      case "keywords":
      case "tags": {
        const keywords = value
          .split(", ")
          .map((k) => k.trim())
          .filter(Boolean);
        if (keywords.length > 0) {
          context.keywords ??= [];
          (context.keywords as string[]).push(...keywords);
        }
        return keywords;
      }
      case "date":
      case "presentation_date": {
        try {
          const presentationDate = new Date(value);
          if (!isNaN(presentationDate.getTime())) {
            context.date = presentationDate;
            return presentationDate;
          }
        } catch {
          // Fall through to default case
        }
        context[underscoredName] = value;
        return value;
      }
      default: {
        // Handle event_* and company_* properties
        const eventMatch = /^(event|company)_(.+)$/.exec(underscoredName);
        if (eventMatch) {
          const [, namespace, propertyName] = eventMatch;
          if (!context[namespace]) {
            context[namespace] = {};
          }
          (context[namespace] as Record<string, unknown>)[propertyName] = value;
          return context[namespace];
        }

        context[underscoredName] = value;
        return value;
      }
    }
  }

  static {
    BaseElement.registerFor.call(this, "meta:user-defined");
  }
}
